### ISSUE TITLE
[Spec] Add #DIFFICULTY to describe song difficulty

### ISSUE THREAD

**Issue by marwin89** [UltraStar-Deluxe/format#34]
### Suggestion

It would be good to have a #DIFFICULTY-tag. It describes how easy or hard a song is to sing. regardless of the difficulty setting "easy - medium - hard" of the pitch-algorithm. Most modern commercial music games use this info.

Song creators estimate the difficulty of a song on a scale of 1-5. Based on tempo, syllables per second, pauses to breath, changes of pitches and octaves. loooong noootes etc.

1 - Very Easy: * - Example: Status Quo - You're in the army now
2 - Easy: ** - Example: ...
3 - Medium: *** - Example: ...
4 - Hard: **** - Example: ...
5 - Very Hard: ***** - Example: ...

```
#DIFFICULTY:5
``` 

### Use case

1. As a user I regullary want to sing easy songs in the beginning to warm up my voice for a karaoke night. 🔥 
2. As a user I want to find easy songs when I'm actually not good at singing or get used to music games. 🎵 
3. Many songs are challenging. Some are very fast and taking your breath away other have complicated pitches to sing.

**Comment by rakuri255** [comment]
Could we define a formula or algorithm? Mayb the program / game could calculate it itself?

**Comment by Nianna** [comment]
I agree with @bohning. If the creator is supposed to define it, it's probably gonna be too subjective to be useful.

Does it have to be a part of the txt file?
It seems like a feature that should be implemented by a website with songfiles (it could be measured by votes) or karaoke game (it could use some algorithm on txt file to determine difficulty).

**Comment by Baklap4** [comment]
Instead of 5 difficulties i'd opt for 3: Easy, Normal, Hard.

Why? A lot of ratings have 5 star rating. The one rating usually picks out 5/5 or 1/5 or 3/5. AKA: It's really good, or really bad or just okay.

 2/5 and 4/5 is rarely picked from my experience since it's a kind of middleground and people tend to go more for absolutes

Aside from integers i'd make it an enum (strings => Easy, Normal, Hard)

**Comment by marwin89** [comment]
> Does it have to be a part of the txt file?
It seems like a feature that should be implemented by a website with songfiles (it could be measured by votes) or karaoke game (it could use some algorithm on txt file to determine difficulty).

I would say, yes it has to be part of the txt file. Why? As a user I want to browse songs by difficulty direct in gameplay not only in usdb. maybe I want to select/filter only easy songs in party mode because I got a night with karaoke newbies. if there's a difficulty tag in txt-files the game doesn't have to calculate the difficulty every time anew on the fly. 

> Yes i'd agree such a thing can be handy, but it pretty subjective as @bohning pointed out.

To make difficulty more objective we should use one algorithm shared in a tool like usdb manager/usdb-syncer/usdb. If that's not possible I would recommend to voting (second). voting is ideally.

To sum it up: regardless of certain subjectiveness, #difficulty comes in handy. it's a progress. I guess a lot of people can often distict between this triangle of easy-medium-hard successfully and update it manually in usdb. So let's put it into v1.1.0 (with recommendation to build an algorithm/use voting).

**Comment by Nianna** [comment]
If you want to store some info in the txt file you have to accept that:
* it will not be updatable (unless user re-downloads the txt)
* many txt files will not have it defined (either because the user already downloaded them or because they use some other website that does not add this tag)

So in a year time maybe, maybe 5% of my collection will have it defined - I have several thousands songs and only a small percentage of them comes from usdb, not everyone uses solely usdb manager/usdb-syncer/usdb.

On the other hand if the algorithm was implemented in the game, then:
 * all songs would have difficulty defined
 * new version of the game could introduce a new algorithm and recalculate difficulties for better accuracy.
 * each game could come up with its own algorithm better than other games.

For the value to be searchable it does not have to be calculated on the fly. It can be calculated once while loading txt files on the game start up (I guess the searchable tags are cached in the memory anyway) or retrieved from a local db and loaded on the start up.

I agree it's a nice feature. 
Just want to point out that by adding it as a tag we are going to limit it to only new songs (and probably only from usdb). Having it in the game would make it work for all songs.

**Comment by marwin89** [comment]
### SUMMARY / VOTE NOW 📈 ✋ 

So here is the summary for this issue. We are close to a final result.

Please vote for this issue till next saturday 14th oct 2023 with emoticon 🎉 / 🚀 / ❤️ .

🚀 A: Don't introduce #DIFFICULTY-tag in txt-file ( but games should calculate them according to an algorithm) to reduce subjectiveness
🎉 B: Introduce #DIFFICUTLY in txt-file (but ideally users vote an average value in usdb, dynamically synced with usdb-syncer) to reduce subjectiveness
❤️ C: Who needs an #DIFFICULTY-tag at all?

What?
#DIFFICULTY describes how easy or hard a song is to sing. do not confuse it with the general difficulty setting "easy - medium - hard" of the game. VALUE: EASY,MEDIUM,HARD (its up to the games how it would be visualised e.g. stars ***). An easy song is slow, has low pitch variations, lots of gaps between notes, short song, etc.

Why these tags?
- As a user I regullary want to sing easy songs in the beginning to warm up my voice for a karaoke night. 🔥
- As a user I want to find easy songs when I'm actually not good at singing or get used to music games. 🎵
- Many songs are challenging. Some are very fast and taking your breath away other have a high pitch-variation.

---

The description in the spec (website) will be something like:

> #DIFFICULTY
> #DIFFICULTY describes how easy or hard a song is to sing. This helps you when you want to warm up or are an unexpierenced singer. Or you explicitly search for challanges. Do not confuse it with the general difficulty setting "easy - medium - hard" of the games. In general an easy song is slow, has low pitch variations, lots of gaps between notes, short song, etc. There are three values: EASY, MEDIUM, HARD.

**Comment by marwin89** [comment]
- I like the idea of an algorithm calculating songdifficulty (Option A) but I doubt that any game will develop such an algorithm in the next years. I would hope for melody mania to do this.
- Thats why I vote (B). To make things clearer I suggest to introduce these values: #DIFFICULTY: Beginner, Advanced, Profi instead of easy, medium, hard (so its harder to confuse it)
- This feature is asked by users I know and in repositories (world party, vocaluxe, etc.). Its IMO hard to explain to the audience why we wouldn't introduce that. It's common in Let's Sing.

**Comment by marwin89** [comment]
### SUMMARY / RESULT 🚩  

Thanks for voting. Here is the final result for this issue.

We do not introduce #DIFFICULTY (option C) for now.  2 🎉 / 5 🚀 + ❤

I may re-open the issue if the overall conditions for development and txt-hosting change.
